src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecBasedSpecProvider.cs

```diff
@@ -203,38 +224,61 @@ private void FillReleaseSpec(ReleaseSpec releaseSpec, long releaseStartBlock, ul
-            releaseSpec.IsEip3651Enabled = (_chainSpec.Parameters.Eip3651TransitionTimestamp ?? ulong.MaxValue) <= releaseStartTimestamp;
+            return releaseSpec;
         }
 
         public void UpdateMergeTransitionInfo(long? blockNumber, UInt256? terminalTotalDifficulty = null)
         {
             if (blockNumber is not null)
-                MergeBlockNumber = blockNumber;
+            {
+                MergeBlockNumber = (ForkActivation)blockNumber;
+            }
+
             if (terminalTotalDifficulty is not null)
+            {
                 TerminalTotalDifficulty = terminalTotalDifficulty;
+            }
         }
 
         public ForkActivation? MergeBlockNumber { get; private set; }
 
         public UInt256? TerminalTotalDifficulty { get; private set; }
 
-        public IReleaseSpec GenesisSpec => _transitions.Length == 0 ? null : _transitions[0].Release;
+        public IReleaseSpec GenesisSpec => _transitions.Length == 0 ? null : _transitions[0].Spec;
+
+        public IReleaseSpec GetSpec(ForkActivation activation)
+        {
+            // TODO: Is this actually needed? Can this be tricked with invalid activation check if someone would fake timestamp from the future?
+            if (_firstTimestampActivation is not null && activation.Timestamp is not null)
+            {
+                if (_firstTimestampActivation.Value.Timestamp < activation.Timestamp
+                    && _firstTimestampActivation.Value.BlockNumber > activation.BlockNumber)
+                {
+                    if (_logger.IsWarn) _logger.Warn($"Chainspec file is misconfigured! Timestamp transition is configured to happen before the last block transition.");
+                }
+            }
+
+            return _transitions.TryGetSearchedItem(activation,
+                CompareTransitionOnActivation,
+                out (ForkActivation Activation, ReleaseSpec Spec) transition)
+                ? transition.Spec
+                : GenesisSpec;
+        }
 
-        public IReleaseSpec GetSpec(ForkActivation forkActivation) =>
-                _transitions.TryGetSearchedItem(forkActivation,
-                    CompareTransitionOnBlock,
-                    out (ForkActivation, ReleaseSpec Release) transition)
-                    ? transition.Release
-                    : null;
+        private static int CompareTransitionOnActivation(ForkActivation activation, (ForkActivation Activation, ReleaseSpec Spec) transition) =>
+            activation.CompareTo(transition.Activation);
 
-        private static int CompareTransitionOnBlock(ForkActivation forkActivation, (ForkActivation, ReleaseSpec Release) transition) =>
-            forkActivation.CompareTo(transition.Item1);
+        private static int CompareTransitionOnTimestamp(ForkActivation activation, (ForkActivation Activation, ReleaseSpec Spec) transition) =>
```

Comment:
@LukaszRozmej this method is not used. Is there a reason for keeping it?

Reply:
We can remove it